lib/semantics/check-omp-structure.cc

```diff
@@ -27,142 +100,168 @@ namespace Fortran::semantics {
 //                        reduction-clause |
 //                        proc-bind-clause
 void OmpStructureChecker::Enter(const parser::OmpBlockDirective::Parallel &x) {
-  std::set<std::type_index> s{typeid(parser::OmpIfClause),
-      typeid(parser::OmpClause::NumThreads), typeid(parser::OmpDefaultClause),
-      typeid(parser::OmpClause::Private),
-      typeid(parser::OmpClause::Firstprivate),
-      typeid(parser::OmpClause::Shared), typeid(parser::OmpClause::Copyin),
-      typeid(parser::OmpReductionClause), typeid(parser::OmpProcBindClause)};
-  SetAllowed(s);
-}
+  if (ompContext_.empty()) {
+    return;
+  }
+  // reserve for nesting check
+  SetOmpContextDirectiveEnum(OmpDirectiveEnum::PARALLEL);
+
+  OmpClauseEnumType allowed{OmpClauseEnum::IF, OmpClauseEnum::NUM_THREADS,
+      OmpClauseEnum::DEFAULT, OmpClauseEnum::PRIVATE,
+      OmpClauseEnum::FIRSTPRIVATE, OmpClauseEnum::SHARED, OmpClauseEnum::COPYIN,
+      OmpClauseEnum::REDUCTION, OmpClauseEnum::PROC_BIND};
+  SetOmpContextAllowed(allowed);
 
-void OmpStructureChecker::Enter(const parser::OmpBlockDirective &x) {
-  SetCurrentDirectiveSource(x.source);
+  OmpClauseEnumType allowedOnce{
+      OmpClauseEnum::IF, OmpClauseEnum::NUM_THREADS, OmpClauseEnum::PROC_BIND};
```

Comment:
Every clause that is allowed once should also be allowed. So you should be able to avoid duplicating clauses between these two sets.

Reply:
nice!

Reply:
This might be reiterating Tim's comment, or it might be more general.  Is it possible to have just a single set for all of these checks, rather than three (allowed, allowed once, seen)?  Initially the set for the current directive is empty.  When a clause is seen, call SetClause( OmpClauseEnum clause).  In SetClause, if the clause is not allowed, generate an error.  If the clause is allowed once, and the clause is already in the set, generate an error.  Then add the clause to the set.  After that, the set is a "seen at least once" set.  Assuming that checks vary by directive, SetClause can contain a switch on the current directive, with each directive case possibly containing a switch on clause, or vice versa, whatever works best.  Or bit sets might be an option vs. switch statements, in which case the switches are replaced with bit indexing or lookup.

Reply:
Addressed in commit [cb89744](cb89744746186bcf120d8ca62a791e2778366b9b). Sorry Val, I didn't follow your suggestion. SetClause may become a giant function.